### ISSUE TITLE
Missing images in Managed WPF UI

### ISSUE THREAD

**Issue by koglerch13** [oleg-shilo/wixsharp#1490]
When using the managed WPF GUI, the Banner and Background images are never shown in my installer. 
I use WixSharp v4, although it seems to affect v3 as well.

![image](https://github.com/oleg-shilo/wixsharp/assets/15924552/2e37152b-55fe-4b29-afaf-0347f19e313f)
![image](https://github.com/oleg-shilo/wixsharp/assets/15924552/aa54277c-1683-434d-9d91-90c53daab52d)

Using the WinForms GUI all the images are shown.

Maybe I have not configured something correctly, but I was unable to find anything in the samples or documentation so far.

That's how I set up the MSI:
```csharp
 project.LicenceFile = @"Assets\License.rtf";
 project.BackgroundImage = @"Images\image_start.bmp";
 project.BannerImage = @"Images\image_banner.bmp";

 project.ManagedUI = new ManagedUI();

 project.ManagedUI.InstallDialogs
     .Add<WelcomeDialog>()
     .Add<LicenceDialog>()
     .Add<CustomDialogWith<ChangelogView>>()
     .Add<FeaturesDialog>()
     .Add<InstallDirDialog>()
     .Add<ProgressDialog>()
     .Add<ExitDialog>();

 project.ManagedUI.ModifyDialogs
     .Add<MaintenanceTypeDialog>()
     .Add<FeaturesDialog>()
     .Add<ProgressDialog>()
     .Add<ExitDialog>();
```

**Comment by koglerch13** [comment]
Thanks for the hint!
If I supply an image with the wrong dimensions the MSI Build fails with this message:
```
'Project.BackgroundImage has incompatible (with ManagedUI default dialogs) aspect ratio. The expected ratio is close to W(156)/H(312). The background image (left side banner) in ManagedUI dialogs is left-docked at runtime and if it's too wide it can push away (to right) the all other UI elements. You can suppress image validation by setting Project.ValidateBackgroundImage to 'false'.'
```

However, In case of my images this does not happen. Instead the MSI builds correctly but the images simply do not show up. 
Even if I use the "default" banner (extracted from the MSI), it does not appear.

This leads me to believe that there is something wrong with the Default WPF GUI.

**Comment by oleg-shilo** [comment]
You can try to debug it.

Create a simple project from the "WixSharp ... Custom UI" template and it will bring all dialog source code. Put an assertion there and then you can debug it.

**Comment by oleg-shilo** [comment]
Read here about how to debug: https://github.com/oleg-shilo/wixsharp/wiki/Tips'n'Tricks#debugging

**Comment by Torchok19081986** [comment]
hm, this one is interesting. What happens if you put your code from above, AFTER last line of  project.ManagedUI.ModifyDialogs ? 

Maybe compiler put it at wrong point of creation ? Only Ole can bring Light to this . 🤔🤔🤔

**Comment by oleg-shilo** [comment]
And another point. If you ever have challenges with rendering or scaling then WPF is a better choice as the technology have the answers to these problems.

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/c9936214-a992-4b27-a614-b81486ab2baf)

**Comment by oleg-shilo** [comment]
@Torchok19081986, no position in the code does not matter.

@koglerch13, if you send me the CustomUI 'hello-world' project with your images I can have a look at it.

**Comment by koglerch13** [comment]
I just debugged a bit and seems like this error is thrown when accessing `session?.GetResourceBitmap("WixSharpUI_Bmp_Dialog").ToImageSource()`.

```
System.Runtime.InteropServices.ExternalException
  HResult=0x80004005
  Message=A generic error occurred in GDI+.
  Source=System.Drawing
  StackTrace:
   at System.Drawing.Image.Save(Stream stream, ImageCodecInfo encoder, EncoderParameters encoderParams)
   at System.Drawing.Image.Save(Stream stream, ImageFormat format)
   at WixSharp.UI.WPF.Extensions.ToImageSource(Bitmap src)
   at WixSharp_Setup1.WelcomeDialogModel.get_Banner() in C:\Development\Neuron.Demonstrator\setup\WixSharp Setup1\Dialogs\WelcomeDialog.xaml.cs:line 50

```

**Comment by koglerch13** [comment]
@oleg-shilo Here's the sample
[WixSharp Setup1.zip](https://github.com/oleg-shilo/wixsharp/files/14866245/WixSharp.Setup1.zip)

**Comment by oleg-shilo** [comment]
>  I am already using WPF, that's where the problem exists.

Which version of Wix are you targeting? And which version of VixSharp?
It was a similar problem with Wix4 as they changed the way embedded background images are stored

**Comment by koglerch13** [comment]
> hm.. I tried also to create msi. Works with your banner and background image. output loos like following 

The first two screens use WinForms, please advance until the 3rd screen.
Sorry, should have noted that of course.

> Which version of Wix are you targeting? And which version of VixSharp?
> It was a similar problem with Wix4 as they changed the way embedded background images are stored.

WiX v4.0.5
WixSharp-wix4 v.2.1.2

**Comment by oleg-shilo** [comment]
Not a problem, will check it tomorrow. It's almost midnight here in AU :)

**Comment by Torchok19081986** [comment]
hallo, @koglerch13. I just tried add your both bmp Files to Template of WPF Wix# V4. Doesnt work either. Somehow project.Banner and project.BackgroundImage wil be ignored. BUT if you compile Template without changes, it shows standard Banner and Images.

**Comment by Torchok19081986** [comment]
update : I just tested your both bmp file into pure wix toolset v4 MSI Template . 

    <WixVariable
  Id="WixUIDialogBmp"
  Value="image_start.bmp"
  />
    <WixVariable
      Id="WixUIBannerBmp"
      Value="image_banner.bmp"
  />

After this works. Now Oleg can search little bit more precise, why in WPF doesnt shows up. 

![Screenshot 2024-04-04 160917](https://github.com/oleg-shilo/wixsharp/assets/25432729/038b953d-caa2-457f-b9d7-bad34c55cbcb)
![Screenshot 2024-04-04 160933](https://github.com/oleg-shilo/wixsharp/assets/25432729/13ecb6d6-5867-4aab-a370-bc3a1a973c60)

**Comment by oleg-shilo** [comment]
I found the problem.

Guys, first let me share the technique for troubleshooting this kind of problem. This is what I did with the code sample @koglerch13 shared:

1. Under debugger I saw that this property throws the exception and the dialog cannot show the image:
   `public BitmapImage Banner => session?.GetResourceBitmap("WixSharpUI_Bmp_Dialog").ToImageSource();`

2. Then I modified the code to show the exact error:
    ![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/6b0d00e3-fd1e-40cd-be83-5a10661416b5)
    This showed me the error at runtime:
    ![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/0dc7758e-206b-4e1c-825e-d0ceee49668c)

This means that for some reason the custom bitmap we provided cannot be rendered by GDI+ runtime for some generic reason. Why the very same routine works for the stock dialogs (first two dialogs) is unknown but probably for the same reason. What this reason exactly, it is a mystery that can only be solved by troubleshooting GDI+. Fortunately, we do not have to.

I just took two properly working PNG images instead of BMP and they started working right away. Meaning that changing bmp to png somehow allows us avoiding this GDI+ problem. Now you can take these images and simply edit then according to your needs. I have attached the working project.

[#1490.zip](https://github.com/oleg-shilo/wixsharp/files/14881761/1490.zip)

**Comment by Torchok19081986** [comment]
good to known. @oleg-shilo can you please change it in sample and also change this issue to enhancement ?

**Comment by oleg-shilo** [comment]
Which sample do you refer to?

**Comment by Torchok19081986** [comment]
custom WPF UI v4 if possible or ad some notice to it into template.

**Comment by oleg-shilo** [comment]
OK. 
Agree. Makes sense.

**Comment by Torchok19081986** [comment]
i have slide suspicions. Somehow is timing is here AFTER security patch of wix toolset. https://github.com/wixtoolset/issues/security/advisories/GHSA-rf39-3f98-xr7r 
Just dont know, if it also somehow affected wix#.

**Comment by oleg-shilo** [comment]
I am not sure it is related to any WiX functionality. It's more like WPF (GDI+) exotic limitations that are easier to work around rather than figure out. 
Thus I will just provide a link to the two working PNG and then it will be up to the user to modify them.

**Comment by oleg-shilo** [comment]
Yep, I also came to the same conclusion but through the experiment. Txs for the link confirming it.

`GetResourceBitmap` uses two memory streams to read from the msi table and then to write to the BitmapImabe. 
somehow Bitmap is not loaded correctly and this triggers the whole problem after that.

Changing the stream to byte array in the `GetResourceBitmap` completely solves the problem and in the direct way.
Will do the release tmr.

This is actually a great outcome. Thank you guys for the help with troubleshooting

**Comment by Torchok19081986** [comment]
null problem. I deal always with Wix toolset no docs Problem and created msi or BA.😉😉😉😉

**Comment by oleg-shilo** [comment]
Done. You can find the fix in the latest v2.1.5 release.

Note that it is a "pure fix" as there is no change in the behaviour but only a change in the stream management when loading the images  from the MSI embedded data: https://github.com/oleg-shilo/wixsharp/commit/e6e21236b13ac18a31bdb26863bb77f3bb7c58ee

**Comment by koglerch13** [comment]
Sorry for the late reply, in v2.1.5 this works perfectly!
Thanks for the help, great work guys!
